File: src/bun/src/installer.ts
/**
 * Installation and update logic
 */

import { join, dirname } from 'path';
import { mkdtemp, rm, readFile, writeFile, cp, mkdir } from 'fs/promises';
import { tmpdir } from 'os';
import { existsSync } from 'fs';
import { spawn } from 'child_process';
import { Downloader } from './downloader';
import { Manifest } from './manifest';

export interface InstallerOptions {
  githubRepo: string;
  verbosity?: number;
}

export class Installer {
  private githubRepo: string;
  private githubRawUrl: string;
  private verbosity: number;
  private downloader: Downloader;

  constructor(options: InstallerOptions) {
    this.githubRepo = options.githubRepo;
    this.githubRawUrl = `https://raw.githubusercontent.com/${options.githubRepo}/main`;
    this.verbosity = options.verbosity ?? 1;
    this.downloader = new Downloader({ verbosity: this.verbosity });
  }

  private logError(message: string): void {
    console.error(`\x1b[31m[ERROR]\x1b[0m ${message}`);
  }

  private logSuccess(message: string): void {
    if (this.verbosity >= 1) {
      console.log(`\x1b[32m[SUCCESS]\x1b[0m ${message}`);
    }
  }

  private logInfo(message: string): void {
    if (this.verbosity >= 1) {
      console.log(`\x1b[34m[INFO]\x1b[0m ${message}`);
    }
  }

  private logWarn(message: string): void {
    if (this.verbosity >= 1) {
      console.log(`\x1b[33m[WARN]\x1b[0m ${message}`);
    }
  }

  private logDebug(message: string): void {
    if (this.verbosity >= 2) {
      console.log(`\x1b[34m[DEBUG]\x1b[0m ${message}`);
    }
  }

  static async parseVersionFile(versionFilePath: string, key: string): Promise<string | null> {
    if (!existsSync(versionFilePath)) {
      return null;
    }

    try {
      const file = Bun.file(versionFilePath);
      const textContent = await file.text();
      const lines = textContent.split('\n');
      
      for (const line of lines) {
        const trimmed = line.trim();
        if (trimmed.startsWith(`${key}=`)) {
          return trimmed.split('=', 2)[1];
        }
      }
      
      return null;
    } catch {
      return null;
    }
  }

  async getRemoteVersion(): Promise<string | null> {
    const url = `${this.githubRawUrl}/.aiassisted/.version`;
    const content = await this.downloader.downloadText(url);

    if (!content) {
      return null;
    }

    // Parse COMMIT_HASH from content
    for (const line of content.split('\n')) {
      const trimmed = line.trim();
      if (trimmed.startsWith('COMMIT_HASH=')) {
        return trimmed.split('=', 2)[1];
      }
    }

    return null;
  }

  async downloadAiassisted(tempDir: string): Promise<boolean> {
    const targetDir = join(tempDir, '.aiassisted');
    await mkdir(targetDir, { recursive: true });

    this.logInfo('Downloading .aiassisted directory...');

    // Download .version file first
    const versionUrl = `${this.githubRawUrl}/.aiassisted/.version`;
    if (!(await this.downloader.downloadFile(versionUrl, join(targetDir, '.version')))) {
      return false;
    }

    // Download FILES.txt manifest
    const manifestUrl = `${this.githubRawUrl}/.aiassisted/FILES.txt`;
    const manifestFile = join(tempDir, 'FILES.txt');
    if (!(await this.downloader.downloadFile(manifestUrl, manifestFile))) {
      this.logError('Failed to download FILES.txt manifest');
      return false;
    }

    // Load manifest
    const manifest = new Manifest({ verbosity: this.verbosity });
    if (!(await manifest.loadFromFile(manifestFile))) {
      return false;
    }

    // Download each file with checksum verification
    for (const [filepath, expectedHash] of manifest.files.entries()) {
      const fileUrl = `${this.githubRawUrl}/.aiassisted/${filepath}`;
      const filePath = join(targetDir, filepath);

      this.logDebug(`Downloading ${filepath}...`);
      if (!(await this.downloader.downloadFile(fileUrl, filePath))) {
        this.logError(`Failed to download ${filepath}`);
        return false;
      }

      // Verify checksum
      if (!(await manifest.verifyChecksum(filePath, expectedHash))) {
        this.logError(`Checksum verification failed for ${filepath}`);
        return false;
      }

      this.logDebug(`Verified checksum for ${filepath}`);
    }

    // Copy FILES.txt to target
    await cp(manifestFile, join(targetDir, 'FILES.txt'));

    this.logSuccess(`Downloaded .aiassisted directory to ${tempDir}`);
    return true;
  }

  async downloadChangedFiles(targetPath: string, remoteManifest: Manifest): Promise<string | null> {
    const localManifestFile = join(targetPath, '.aiassisted', 'FILES.txt');
    const localManifest = new Manifest({ verbosity: this.verbosity });

    if (existsSync(localManifestFile)) {
      await localManifest.loadFromFile(localManifestFile);
    }

    // Compare manifests
    const { changed, unchanged } = localManifest.compareWith(remoteManifest);

    this.logInfo(`Changed: ${changed.length} file(s), Unchanged: ${unchanged.length} file(s)`);

    // Create temp directory
    const tempDir = await mkdtemp(join(tmpdir(), 'aiassisted-'));
    const targetDir = join(tempDir, '.aiassisted');
    await mkdir(targetDir, { recursive: true });

    // Download only changed files
    for (const filepath of changed) {
      const expectedHash = remoteManifest.files.get(filepath)!;
      const fileUrl = `${this.githubRawUrl}/.aiassisted/${filepath}`;
      const filePath = join(targetDir, filepath);

      this.logDebug(`Downloading changed file: ${filepath}...`);
      if (!(await this.downloader.downloadFile(fileUrl, filePath))) {
        this.logError(`Failed to download ${filepath}`);
        await rm(tempDir, { recursive: true, force: true });
        return null;
      }

      // Verify checksum
      if (!(await remoteManifest.verifyChecksum(filePath, expectedHash))) {
        this.logError(`Checksum verification failed for ${filepath}`);
        await rm(tempDir, { recursive: true, force: true });
        return null;
      }
    }

    return tempDir;
  }

  async install(targetPath: string): Promise<boolean> {
    // Create temp directory
    const tempDir = await mkdtemp(join(tmpdir(), 'aiassisted-'));

    try {
      // Download .aiassisted
      if (!(await this.downloadAiassisted(tempDir))) {
        return false;
      }

      // Copy to target
      const targetDir = join(targetPath, '.aiassisted');
      if (existsSync(targetDir)) {
        await rm(targetDir, { recursive: true, force: true });
      }

      await cp(join(tempDir, '.aiassisted'), targetDir, { recursive: true });

      this.logSuccess(`Installed .aiassisted to ${targetPath}`);
      return true;
    } finally {
      await rm(tempDir, { recursive: true, force: true });
    }
  }

  async update(targetPath: string, force: boolean = false): Promise<boolean> {
    // Download remote manifest
    const manifestUrl = `${this.githubRawUrl}/.aiassisted/FILES.txt`;
    const manifestContent = await this.downloader.downloadText(manifestUrl);

    if (!manifestContent) {
      this.logError('Failed to download remote manifest');
      return false;
    }

    const remoteManifest = new Manifest({ verbosity: this.verbosity });
    if (!remoteManifest.loadFromContent(manifestContent)) {
      return false;
    }

    // Download only changed files
    const tempDir = await this.downloadChangedFiles(targetPath, remoteManifest);
    if (!tempDir) {
      return false;
    }

    try {
      // Download .version file
      const versionUrl = `${this.githubRawUrl}/.aiassisted/.version`;
      if (!(await this.downloader.downloadFile(versionUrl, join(tempDir, '.aiassisted', '.version')))) {
        this.logError('Failed to download .version file');
        return false;
      }

      // Show diff unless --force
      if (!force) {
        const tempAiassisted = join(tempDir, '.aiassisted');
        const hasChanges = existsSync(tempAiassisted);

        if (hasChanges) {
          console.log('\n\x1b[1m\x1b[33mChanges to be applied:\x1b[0m\n');

          // Generate diff
          try {
            const diff = await this.generateDiff(
              join(targetPath, '.aiassisted'),
              tempAiassisted
            );
            
            if (diff) {
              const lines = diff.split('\n');
              for (const line of lines.slice(0, 100)) {
                console.log(line);
              }

              if (lines.length > 100) {
                console.log(`\n\x1b[34m... (${lines.length - 100} more lines)\x1b[0m\n`);
              } else {
                console.log();
              }
            }
          } catch {
            this.logWarn('Could not generate diff');
          }
        } else {
          console.log('\n\x1b[1m\x1b[34mNo file content changes (version metadata update only)\x1b[0m\n');
        }

        // Prompt for confirmation
        const answer = prompt('Apply these changes? [y/N]: ');
        if (!answer || !['y', 'yes'].includes(answer.toLowerCase())) {
          this.logInfo('Update cancelled');
          return false;
        }
      }

      // Apply selective updates
      if (!(await this.applySelectiveUpdate(tempDir, targetPath, remoteManifest))) {
        return false;
      }

      return true;
    } finally {
      await rm(tempDir, { recursive: true, force: true });
    }
  }

  private async generateDiff(oldPath: string, newPath: string): Promise<string | null> {
    return new Promise((resolve) => {
      const diff = spawn('diff', ['-ru', oldPath, newPath]);
      let output = '';

      diff.stdout.on('data', (data) => {
        output += data.toString();
      });

      diff.on('close', () => {
        resolve(output || null);
      });

      diff.on('error', () => {
        resolve(null);
      });
    });
  }

  private async applySelectiveUpdate(
    sourceDir: string,
    targetPath: string,
    remoteManifest: Manifest
  ): Promise<boolean> {
    this.logInfo('Applying selective updates...');

    const targetAiassisted = join(targetPath, '.aiassisted');
    const sourceAiassisted = join(sourceDir, '.aiassisted');

    let updated = 0;
    let skipped = 0;

    // Update .version file
    const versionSource = join(sourceAiassisted, '.version');
    const versionTarget = join(targetAiassisted, '.version');
    if (existsSync(versionSource)) {
      await cp(versionSource, versionTarget);
    }

    // Update FILES.txt manifest
    const manifestTarget = join(targetAiassisted, 'FILES.txt');
    let manifestContent = '';
    for (const [filepath, hash] of remoteManifest.files.entries()) {
      manifestContent += `${filepath}:${hash}\n`;
    }
    await writeFile(manifestTarget, manifestContent);

    // Copy only changed files
    for (const filepath of remoteManifest.files.keys()) {
      const sourceFile = join(sourceAiassisted, filepath);
      const targetFile = join(targetAiassisted, filepath);

      // If file exists in source (was downloaded), copy it
      if (existsSync(sourceFile)) {
        await mkdir(dirname(targetFile), { recursive: true });
        await cp(sourceFile, targetFile);
        this.logDebug(`Updated: ${filepath}`);
        updated++;
      } else {
        skipped++;
      }
    }

    this.logSuccess(`Updated ${updated} file(s), skipped ${skipped} unchanged file(s)`);
    return true;
  }
}

File: src/python/aiassisted/installer.py
"""Installation and update logic"""

import shutil
import tempfile
import subprocess
from pathlib import Path
from typing import Optional
from rich.console import Console
from rich.prompt import Confirm
from .downloader import Downloader
from .manifest import Manifest

console = Console()


class Installer:
    """Handle installation and updates of .aiassisted directory"""

    def __init__(self, github_repo: str, verbosity: int = 1):
        self.github_repo = github_repo
        self.github_raw_url = f"https://raw.githubusercontent.com/{github_repo}/main"
        self.verbosity = verbosity
        self.downloader = Downloader(verbosity=verbosity)

    def log_error(self, message: str):
        """Log error message"""
        console.print(f"[bold red][ERROR][/bold red] {message}", file=sys.stderr)

    def log_success(self, message: str):
        """Log success message"""
        if self.verbosity >= 1:
            console.print(f"[bold green][SUCCESS][/bold green] {message}")

    def log_info(self, message: str):
        """Log info message"""
        if self.verbosity >= 1:
            console.print(f"[bold blue][INFO][/bold blue] {message}")

    def log_warn(self, message: str):
        """Log warning message"""
        if self.verbosity >= 1:
            console.print(f"[bold yellow][WARN][/bold yellow] {message}")

    def log_debug(self, message: str):
        """Log debug message"""
        if self.verbosity >= 2:
            console.print(f"[blue][DEBUG][/blue] {message}")

    @staticmethod
    def parse_version_file(version_file: Path, key: str) -> Optional[str]:
        """Parse version file for specific key

        Args:
            version_file: Path to .version file
            key: Key to extract (e.g., COMMIT_HASH)

        Returns:
            Value if found, None otherwise
        """
        if not version_file.exists():
            return None

        try:
            with open(version_file, "r") as f:
                for line in f:
                    line = line.strip()
                    if line.startswith(f"{key}="):
                        return line.split("=", 1)[1]
            return None
        except Exception:
            return None

    def get_remote_version(self) -> Optional[str]:
        """Fetch remote version hash

        Returns:
            Remote COMMIT_HASH or None on failure
        """
        url = f"{self.github_raw_url}/.aiassisted/.version"
        content = self.downloader.download_text(url)

        if content is None:
            return None

        # Parse COMMIT_HASH from content
        for line in content.split("\n"):
            line = line.strip()
            if line.startswith("COMMIT_HASH="):
                return line.split("=", 1)[1]

        return None

    def download_aiassisted(self, temp_dir: Path) -> bool:
        """Download complete .aiassisted directory to temp location

        Args:
            temp_dir: Temporary directory to download to

        Returns:
            True on success, False on failure
        """
        target_dir = temp_dir / ".aiassisted"
        target_dir.mkdir(parents=True, exist_ok=True)

        self.log_info("Downloading .aiassisted directory...")

        # Download .version file first
        version_url = f"{self.github_raw_url}/.aiassisted/.version"
        if not self.downloader.download_file(version_url, target_dir / ".version"):
            return False

        # Download FILES.txt manifest
        manifest_url = f"{self.github_raw_url}/.aiassisted/FILES.txt"
        manifest_file = temp_dir / "FILES.txt"
        if not self.downloader.download_file(manifest_url, manifest_file):
            self.log_error("Failed to download FILES.txt manifest")
            return False

        # Load manifest
        manifest = Manifest(verbosity=self.verbosity)
        if not manifest.load_from_file(manifest_file):
            return False

        # Download each file with checksum verification
        for filepath, expected_hash in manifest.files.items():
            file_url = f"{self.github_raw_url}/.aiassisted/{filepath}"
            file_path = target_dir / filepath

            self.log_debug(f"Downloading {filepath}...")
            if not self.downloader.download_file(file_url, file_path):
                self.log_error(f"Failed to download {filepath}")
                return False

            # Verify checksum
            if not manifest.verify_checksum(file_path, expected_hash):
                self.log_error(f"Checksum verification failed for {filepath}")
                return False

            self.log_debug(f"Verified checksum for {filepath}")

        # Copy FILES.txt to target
        shutil.copy(manifest_file, target_dir / "FILES.txt")

        self.log_success(f"Downloaded .aiassisted directory to {temp_dir}")
        return True

    def download_changed_files(
        self, target_path: Path, remote_manifest: Manifest
    ) -> Optional[Path]:
        """Download only changed files based on checksums

        Args:
            target_path: Target installation path
            remote_manifest: Remote manifest to compare with

        Returns:
            Path to temp directory with changed files, or None on failure
        """
        local_manifest_file = target_path / ".aiassisted" / "FILES.txt"
        local_manifest = Manifest(verbosity=self.verbosity)

        if local_manifest_file.exists():
            local_manifest.load_from_file(local_manifest_file)

        # Compare manifests
        changed, unchanged = local_manifest.compare_with(remote_manifest)

        self.log_info(
            f"Changed: {len(changed)} file(s), Unchanged: {len(unchanged)} file(s)"
        )

        # Create temp directory
        temp_dir = Path(tempfile.mkdtemp())
        target_dir = temp_dir / ".aiassisted"
        target_dir.mkdir(parents=True, exist_ok=True)

        # Download only changed files
        for filepath in changed:
            expected_hash = remote_manifest.files[filepath]
            file_url = f"{self.github_raw_url}/.aiassisted/{filepath}"
            file_path = target_dir / filepath

            self.log_debug(f"Downloading changed file: {filepath}...")
            if not self.downloader.download_file(file_url, file_path):
                self.log_error(f"Failed to download {filepath}")
                shutil.rmtree(temp_dir, ignore_errors=True)
                return None

            # Verify checksum
            if not remote_manifest.verify_checksum(file_path, expected_hash):
                self.log_error(f"Checksum verification failed for {filepath}")
                shutil.rmtree(temp_dir, ignore_errors=True)
                return None

        return temp_dir

    def install(self, target_path: Path) -> bool:
        """Install .aiassisted directory

        Args:
            target_path: Directory to install to

        Returns:
            True on success, False on failure
        """
        # Create temp directory
        temp_dir = Path(tempfile.mkdtemp())

        try:
            # Download .aiassisted
            if not self.download_aiassisted(temp_dir):
                return False

            # Copy to target
            target_dir = target_path / ".aiassisted"
            if target_dir.exists():
                shutil.rmtree(target_dir)

            shutil.copytree(temp_dir / ".aiassisted", target_dir)

            self.log_success(f"Installed .aiassisted to {target_path}")
            return True
        finally:
            shutil.rmtree(temp_dir, ignore_errors=True)

    def update(self, target_path: Path, force: bool = False) -> bool:
        """Update existing .aiassisted installation

        Args:
            target_path: Directory containing .aiassisted
            force: Skip confirmation prompts

        Returns:
            True on success, False on failure
        """
        # Download remote manifest
        manifest_url = f"{self.github_raw_url}/.aiassisted/FILES.txt"
        manifest_content = self.downloader.download_text(manifest_url)

        if manifest_content is None:
            self.log_error("Failed to download remote manifest")
            return False

        remote_manifest = Manifest(verbosity=self.verbosity)
        if not remote_manifest.load_from_content(manifest_content):
            return False

        # Download only changed files
        temp_dir = self.download_changed_files(target_path, remote_manifest)
        if temp_dir is None:
            return False

        try:
            # Download .version file
            version_url = f"{self.github_raw_url}/.aiassisted/.version"
            if not self.downloader.download_file(
                version_url, temp_dir / ".aiassisted" / ".version"
            ):
                self.log_error("Failed to download .version file")
                return False

            # Show diff unless --force
            if not force:
                has_changes = any((temp_dir / ".aiassisted").iterdir())

                if has_changes:
                    console.print(
                        "\n[bold yellow]Changes to be applied:[/bold yellow]\n"
                    )

                    # Generate diff
                    try:
                        result = subprocess.run(
                            [
                                "diff",
                                "-ru",
                                str(target_path / ".aiassisted"),
                                str(temp_dir / ".aiassisted"),
                            ],
                            capture_output=True,
                            text=True,
                        )

                        diff_output = result.stdout
                        if diff_output:
                            lines = diff_output.split("\n")
                            for line in lines[:100]:
                                console.print(line)

                            if len(lines) > 100:
                                console.print(
                                    f"\n[blue]... ({len(lines) - 100} more lines)[/blue]\n"
                                )
                            else:
                                console.print()
                    except Exception:
                        self.log_warn("Could not generate diff")
                else:
                    console.print(
                        "\n[bold blue]No file content changes (version metadata update only)[/bold blue]\n"
                    )

                # Prompt for confirmation
                if not Confirm.ask("Apply these changes?"):
                    self.log_info("Update cancelled")
                    return False

            # Apply selective updates
            if not self.apply_selective_update(temp_dir, target_path, remote_manifest):
                return False

            return True
        finally:
            shutil.rmtree(temp_dir, ignore_errors=True)

    def apply_selective_update(
        self, source_dir: Path, target_path: Path, remote_manifest: Manifest
    ) -> bool:
        """Apply selective updates (only changed files)

        Args:
            source_dir: Temp directory with downloaded changes
            target_path: Target installation path
            remote_manifest: Remote manifest

        Returns:
            True on success, False on failure
        """
        self.log_info("Applying selective updates...")

        target_aiassisted = target_path / ".aiassisted"
        source_aiassisted = source_dir / ".aiassisted"

        updated = 0
        skipped = 0

        # Update .version file
        version_source = source_aiassisted / ".version"
        version_target = target_aiassisted / ".version"
        if version_source.exists():
            shutil.copy(version_source, version_target)

        # Update FILES.txt manifest - write from remote_manifest
        manifest_target = target_aiassisted / "FILES.txt"
        with open(manifest_target, "w") as f:
            for filepath, hash_value in remote_manifest.files.items():
                f.write(f"{filepath}:{hash_value}\n")

        # Copy only changed files
        for filepath in remote_manifest.files.keys():
            source_file = source_aiassisted / filepath
            target_file = target_aiassisted / filepath

            # If file exists in source (was downloaded), copy it
            if source_file.exists():
                target_file.parent.mkdir(parents=True, exist_ok=True)
                shutil.copy(source_file, target_file)
                self.log_debug(f"Updated: {filepath}")
                updated += 1
            else:
                skipped += 1

        self.log_success(
            f"Updated {updated} file(s), skipped {skipped} unchanged file(s)"
        )
        return True


import sys
